### PR TITLE
[CALCITE-6663] Support SPLIT_PART function for PostgreSql

### PR DESCRIPTION
**Pull request by zhuqi-lucas**
We can support split_part function for postgresq:

split_part ( string text, delimiter text, n integer ) → text

Splits string at occurrences of delimiter and returns the n'th field (counting from one), or when n is negative, returns the |n|'th-from-last field.
```
split_part('abc@~def@~ghi', '@~', 2) → def

split_part('abc,def,ghi,jkl', ',', -2) → ghi
```

### CODE REVIEW

**Comment by NobiGo** [comment]
> We can support split_part function for postgresq:
> 
> split_part ( string text, delimiter text, n integer ) → text
> 
> Splits string at occurrences of delimiter and returns the n'th field (counting from one), or when n is negative, returns the |n|'th-from-last field.
> 
> ```
> split_part('abc~@def@ghi', '@~', 2) → def
> 
> split_part('abc,def,ghi,jkl', ',', -2) → ghi
> ```

select split_part('abc~@def@ghi', '@~', 2); return nothing in PG. There is a typographical error here.

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java
Comment (NobiGo) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:10907]: There should be three parameters?

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java
Comment (NobiGo) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:10904]: Add unit test to cover `REDSHIFT`? I notice this function is except for `REDSHIFT`.

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java
Comment (NobiGo) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:10912]: Do we should return `VARCHAR(15)` for this function? I want to confirm is there a document explaining？

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java
Comment (NobiGo) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:10921]: Add unit test to cover NULL parameter.

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java
Comment (mihaibudiu) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:10912]: This type comes from ARG0_NULLABLE.
It is true that the result will never be larger.
Another reasonable type would be VARCHAR.
I think there are many functions which use this result type inference.
But indeed, the VARCHAR(N) types have some unintuitive behaviors sometimes.

**Comment by zhuqi-lucas** [comment]
> > We can support split_part function for postgresq:
> > split_part ( string text, delimiter text, n integer ) → text
> > Splits string at occurrences of delimiter and returns the n'th field (counting from one), or when n is negative, returns the |n|'th-from-last field.
> > ```
> > split_part('abc~@def@ghi', '@~', 2) → def
> > 
> > split_part('abc,def,ghi,jkl', ',', -2) → ghi
> > ```
> 
> select split_part('abc~@def@ghi', '@~', 2); return nothing in PG. There is a typographical error here.

Updated, thanks!

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java
Comment (zhuqi-lucas) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: Thank you @NobiGo for review, addressed in latest PR.

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java
Comment (zhuqi-lucas) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: Addressed in latest PR, thanks!

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java
Comment (zhuqi-lucas) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java]: Addressed in latest PR, thanks!

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java
Comment (zhuqi-lucas) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:10904]: Added the test, addressed in latest PR, thanks!

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java
Comment (zhuqi-lucas) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java]: Changed to VARCHAR , addressed in latest PR, thanks!

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java
Comment (zhuqi-lucas) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java]: Addressed in latest PR, thanks!

Comment (NobiGo) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java]: I noticed you did not covered the `SPLIT_PART('abc,,ghi', ',', null)`

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java
Comment (zhuqi-lucas) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: Good catch, addressed in latest PR.

Comment (zhuqi-lucas) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: Addressed in latest PR, thanks!

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java
Comment (zhuqi-lucas) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java]: Added now the third paras null case, thanks!

**Review by NobiGo — APPROVED** [review]
LGTM.

**Comment by mihaibudiu** [comment]
This has already two approvals, Can you please sqash the commits to a single one, which should have the same title as the JIRA issue? Then we can merge the PR. Thank you.

**Comment by zhuqi-lucas** [comment]
> This has already two approvals, Can you please sqash the commits to a single one, which should have the same title as the JIRA issue? Then we can merge the PR. Thank you.

Done, thanks @mihaibudiu !

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java
Comment (caicancai) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:10904]: Minor: Add jira link

Comment (zhuqi-lucas) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java]: Added now, thanks!

**Comment by mihaibudiu** [comment]
Will merge this after the CI completes

**Pull request merged**
